src/serac/physics/heat_transfer.hpp

```diff
@@ -237,34 +262,28 @@ class HeatTransfer<order, dim, Parameters<parameter_space...>, std::integer_sequ
   template <int... active_parameters, typename MaterialType>
   void setMaterial(DependsOn<active_parameters...>, MaterialType material)
   {
-    K_functional_->AddDomainIntegral(
-        Dimension<dim>{}, DependsOn<0, active_parameters + 1 ...>{},
-        [material](auto x, auto temperature, auto... params) {
+    residual_->AddDomainIntegral(
+        Dimension<dim>{}, DependsOn<0, 1, 2, active_parameters + NUM_STATE_VARS...>{},
+        [material](auto x, auto temperature, auto dtemp_dt, auto shape, auto... params) {
           // Get the value and the gradient from the input tuple
-          auto [u, du_dx] = temperature;
-          auto source     = serac::zero{};
-
-          auto response = material(x, u, du_dx, params...);
+          auto [u, du_dX] = temperature;
+          auto [p, dp_dX] = shape;
+          auto du_dt      = get<VALUE>(dtemp_dt);
 
-          return serac::tuple{source, -1.0 * response.heat_flux};
-        },
-        mesh_);
-
-    M_functional_->AddDomainIntegral(
-        Dimension<dim>{}, DependsOn<0, active_parameters + 1 ...>{},
-        [material](auto x, auto d_temperature_dt, auto... params) {
-          auto [u, du_dx] = d_temperature_dt;
-          auto flux       = serac::zero{};
-
-          auto temp      = u * 0.0;
-          auto temp_grad = du_dx * 0.0;
+          // Note that the current configuration x = X + p, where X is the original reference
+          // configuration and p is the shape displacement. We need the gradient with
+          // respect to the perturbed reference configuration x = X + p for the material model. Therefore, we calculate
+          // du/dx = du/dX * dX/dx = du/dX * (dx/dX)^-1 = du/dX * (I + dp/dX)^-1
 
-          auto response = material(x, temp, temp_grad, params...);
+          auto du_dx = dot(du_dX, inv(I + dp_dX));
 
-          auto source = response.specific_heat_capacity * response.density * u;
+          auto response = material(x + p, u, du_dx, params...);
+          auto source   = response.specific_heat_capacity * response.density * du_dt;
```

Comment:
`density` appears in the response struct and also the material type as a member variable.

Would it be appropriate to remove `density` from the response and just get it through the material model? (this is what we're doing in the solid mechanics module)

Reply:
I just changed the thermal material interface to look more like the solid one (i.e. the `()` call returns flux and heat capacity and density are public struct members). I think further iterations on this interface should involve both modules to ensure they are consistent.

Reply:
I think density w.r.t. reference volume makes sense for a public member variable (i.e. it's homogeneous within a material block), but isn't heat capacity is a function of temperature and other stuff (@btalamini knows better than I do)?

Reply:
Yes, it's a response function, and can't be assumed constant in general.

Reply:
> another possible option is to make the thermal materials responsible for outputting the "source" term directly (which would require them to take in the time rate of the temperature, too).

I agree, there is an argument to be made for this. It would put the time and space derivatives on an equal footing in the q-function signature. Howver, it's hard for me to see if one way is better than the other yet. I'm also ok with making the heat capacity an output of the q-function.

Reply:
I think I'm leaning toward outputting the heat capacity

Reply:
I'll take a stab at this change.

Reply:
Now the thermal material returns a serac tuple of volumetric heat capacity and thermal flux directly. Let me know what you think!